File: prompts/loader.py
"""
Universal Prompt Loader for Python

Load prompts from:
1. Supabase database (highest priority)
2. Local YAML files (fallback)

Usage:
    from prompts.loader import PromptLoader
    
    loader = PromptLoader()
    loader.with_supabase(os.environ['SUPABASE_URL'], os.environ['SUPABASE_KEY'])
    
    prompt = loader.get('nko_frame')
    filled = loader.fill(prompt.template, {'nko_text': 'ߒߞߏ', 'translation': 'I declare'})
"""

import os
import re
import yaml
import httpx
from pathlib import Path
from dataclasses import dataclass, field
from typing import Optional, Dict, List, Any


@dataclass
class PromptDefinition:
    """A prompt definition loaded from database or file."""
    id: str
    name: str
    template: str
    category: str
    cost_tier: str = "text_only"  # text_only, multimodal, live
    description: Optional[str] = None
    output_schema: Optional[Dict[str, Any]] = None
    version: str = "1.0.0"
    active: bool = True
    tags: List[str] = field(default_factory=list)


class PromptLoader:
    """
    Universal prompt loader with Supabase and YAML support.
    
    Priority:
    1. Supabase database (if configured)
    2. Local YAML files
    """
    
    def __init__(self, yaml_dir: Optional[str] = None):
        """
        Initialize the prompt loader.
        
        Args:
            yaml_dir: Directory containing YAML prompt files.
                     Defaults to 'nko' subdirectory.
        """
        if yaml_dir:
            self.yaml_dir = Path(yaml_dir)
        else:
            self.yaml_dir = Path(__file__).parent / "nko"
        
        self.supabase_url: Optional[str] = None
        self.supabase_key: Optional[str] = None
        self._cache: Dict[str, PromptDefinition] = {}
    
    def with_supabase(self, url: str, key: str) -> "PromptLoader":
        """Configure Supabase connection."""
        self.supabase_url = url
        self.supabase_key = key
        return self
    
    def from_env(self) -> "PromptLoader":
        """Configure from environment variables."""
        if url := os.environ.get("SUPABASE_URL"):
            if key := os.environ.get("SUPABASE_SERVICE_KEY") or os.environ.get("SUPABASE_ANON_KEY"):
                self.with_supabase(url, key)
        return self
    
    def get(self, id: str) -> Optional[PromptDefinition]:
        """
        Get a prompt by ID.
        
        Priority: Cache → Supabase → YAML
        """
        # Check cache
        if id in self._cache:
            return self._cache[id]
        
        # Try Supabase
        if self.supabase_url and self.supabase_key:
            prompt = self._load_from_supabase(id)
            if prompt:
                self._cache[id] = prompt
                return prompt
        
        # Fall back to YAML
        prompt = self._load_from_yaml(id)
        if prompt:
            self._cache[id] = prompt
        return prompt
    
    def get_template(self, id: str) -> Optional[str]:
        """Get just the template string."""
        prompt = self.get(id)
        return prompt.template if prompt else None
    
    def fill(self, template: str, variables: Dict[str, str]) -> str:
        """
        Fill template placeholders with values.
        
        Placeholders use {variable_name} syntax.
        """
        result = template
        for key, value in variables.items():
            result = result.replace(f"{{{key}}}", value)
        return result
    
    def get_and_fill(self, id: str, variables: Dict[str, str]) -> Optional[str]:
        """Get a prompt and fill it with variables in one call."""
        template = self.get_template(id)
        if not template:
            return None
        return self.fill(template, variables)
    
    def list_ids(self) -> List[str]:
        """List all available prompt IDs."""
        ids = set()
        
        # From Supabase
        if self.supabase_url and self.supabase_key:
            try:
                with httpx.Client() as client:
                    response = client.get(
                        f"{self.supabase_url}/rest/v1/nko_prompts",
                        params={"select": "id", "active": "eq.true"},
                        headers={
                            "apikey": self.supabase_key,
                            "Authorization": f"Bearer {self.supabase_key}",
                        },
                    )
                    if response.is_success:
                        for row in response.json():
                            ids.add(row["id"])
            except Exception as e:
                print(f"Warning: Failed to list prompts from Supabase: {e}")
        
        # From YAML
        ids.update(self._load_all_yaml_ids())
        
        return sorted(ids)
    
    def list_by_category(self, category: str) -> List[PromptDefinition]:
        """List prompts by category."""
        prompts = []
        for id in self.list_ids():
            prompt = self.get(id)
            if prompt and prompt.category == category:
                prompts.append(prompt)
        return prompts
    
    def get_world_prompts(self) -> List[PromptDefinition]:
        """Get all world exploration prompts."""
        return self.list_by_category("world_exploration")
    
    def get_frame_prompts(self) -> List[PromptDefinition]:
        """Get all frame analysis prompts."""
        return self.list_by_category("frame_analysis")
    
    def clear_cache(self) -> None:
        """Clear the prompt cache."""
        self._cache.clear()
    
    # Private methods
    
    def _load_from_supabase(self, id: str) -> Optional[PromptDefinition]:
        """Load a prompt from Supabase."""
        if not self.supabase_url or not self.supabase_key:
            return None
        
        try:
            with httpx.Client() as client:
                response = client.get(
                    f"{self.supabase_url}/rest/v1/nko_prompts",
                    params={"id": f"eq.{id}", "active": "eq.true"},
                    headers={
                        "apikey": self.supabase_key,
                        "Authorization": f"Bearer {self.supabase_key}",
                    },
                )
                
                if not response.is_success:
                    return None
                
                rows = response.json()
                if not rows:
                    return None
                
                row = rows[0]
                return PromptDefinition(
                    id=row["id"],
                    name=row["name"],
                    description=row.get("description"),
                    category=row["category"],
                    cost_tier=row.get("cost_tier", "text_only"),
                    template=row["template"],
                    output_schema=row.get("output_schema"),
                    version=row.get("version", "1.0.0"),
                    active=row.get("active", True),
                    tags=row.get("tags", []),
                )
        except Exception as e:
            print(f"Warning: Failed to load prompt {id} from Supabase: {e}")
            return None
    
    def _load_from_yaml(self, id: str) -> Optional[PromptDefinition]:
        """Load a prompt from YAML files."""
        for file_path in self._get_yaml_files():
            try:
                with open(file_path) as f:
                    data = yaml.safe_load(f)
                
                if "prompts" in data and id in data["prompts"]:
                    prompt_data = data["prompts"][id]
                    return PromptDefinition(
                        id=prompt_data.get("id", id),
                        name=prompt_data.get("name", id),
                        description=prompt_data.get("description"),
                        category=prompt_data.get("category", data.get("category", "custom")),
                        cost_tier=prompt_data.get("cost_tier", data.get("cost_tier", "text_only")),
                        template=prompt_data["template"],
                        output_schema=prompt_data.get("output_schema"),
                        version=prompt_data.get("version", "1.0.0"),
                        active=prompt_data.get("active", True),
                        tags=prompt_data.get("tags", []),
                    )
            except Exception as e:
                print(f"Warning: Failed to load YAML file {file_path}: {e}")
        
        return None
    
    def _load_all_yaml_ids(self) -> List[str]:
        """Load all prompt IDs from YAML files."""
        ids = []
        for file_path in self._get_yaml_files():
            try:
                with open(file_path) as f:
                    data = yaml.safe_load(f)
                if "prompts" in data:
                    ids.extend(data["prompts"].keys())
            except Exception as e:
                print(f"Warning: Failed to parse YAML file {file_path}: {e}")
        return ids
    
    def _get_yaml_files(self) -> List[Path]:
        """Get all YAML files in the yaml_dir."""
        files = []
        if self.yaml_dir.exists():
            for entry in self.yaml_dir.iterdir():
                if entry.is_file() and entry.suffix in (".yaml", ".yml"):
                    files.append(entry)
        return files


# Singleton instance
_loader: Optional[PromptLoader] = None


def get_loader() -> PromptLoader:
    """Get the singleton prompt loader instance."""
    global _loader
    if _loader is None:
        _loader = PromptLoader().from_env()
    return _loader


def get_prompt(id: str) -> Optional[str]:
    """Get a prompt template by ID."""
    return get_loader().get_template(id)


def get_and_fill_prompt(id: str, variables: Dict[str, str]) -> Optional[str]:
    """Get and fill a prompt in one call."""
    return get_loader().get_and_fill(id, variables)


# CLI for testing
if __name__ == "__main__":
    import sys
    
    loader = PromptLoader().from_env()
    
    if len(sys.argv) < 2:
        print("Usage: python loader.py <prompt_id> [var1=val1 var2=val2 ...]")
        print("\nAvailable prompts:")
        for id in loader.list_ids():
            print(f"  - {id}")
        sys.exit(0)
    
    prompt_id = sys.argv[1]
    variables = {}
    
    for arg in sys.argv[2:]:
        if "=" in arg:
            key, value = arg.split("=", 1)
            variables[key] = value
    
    if variables:
        result = loader.get_and_fill(prompt_id, variables)
    else:
        result = loader.get_template(prompt_id)
    
    if result:
        print(result)
    else:
        print(f"Prompt '{prompt_id}' not found")
        sys.exit(1)

File: prompts/loader.ts
/**
 * Universal Prompt Loader
 * 
 * Load prompts from:
 * 1. Supabase database (highest priority)
 * 2. Local YAML files (fallback)
 * 
 * Usage:
 * ```typescript
 * import { PromptLoader } from './prompts/loader';
 * 
 * const loader = new PromptLoader()
 *   .withSupabase(process.env.SUPABASE_URL!, process.env.SUPABASE_KEY!);
 * 
 * const prompt = await loader.get('nko_frame');
 * const filled = loader.fill(prompt, { nko_text: 'ߒߞߏ', translation: 'I declare' });
 * ```
 */

import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';

export interface PromptDefinition {
  id: string;
  name: string;
  description?: string;
  category: string;
  cost_tier: 'text_only' | 'multimodal' | 'live';
  template: string;
  output_schema?: Record<string, unknown>;
  version?: string;
  active?: boolean;
  tags?: string[];
}

export interface PromptFile {
  schema_version: string;
  category: string;
  cost_tier: string;
  prompts: Record<string, PromptDefinition>;
}

export class PromptLoader {
  private supabaseUrl?: string;
  private supabaseKey?: string;
  private cache: Map<string, PromptDefinition> = new Map();
  private yamlDir: string;

  constructor(yamlDir?: string) {
    this.yamlDir = yamlDir || path.join(__dirname, 'nko');
  }

  /**
   * Configure Supabase connection
   */
  withSupabase(url: string, key: string): this {
    this.supabaseUrl = url;
    this.supabaseKey = key;
    return this;
  }

  /**
   * Get a prompt by ID
   */
  async get(id: string): Promise<PromptDefinition | null> {
    // Check cache
    if (this.cache.has(id)) {
      return this.cache.get(id)!;
    }

    // Try Supabase first
    if (this.supabaseUrl && this.supabaseKey) {
      const prompt = await this.loadFromSupabase(id);
      if (prompt) {
        this.cache.set(id, prompt);
        return prompt;
      }
    }

    // Fall back to YAML
    const prompt = this.loadFromYaml(id);
    if (prompt) {
      this.cache.set(id, prompt);
    }
    return prompt;
  }

  /**
   * Get prompt template content
   */
  async getTemplate(id: string): Promise<string | null> {
    const prompt = await this.get(id);
    return prompt?.template || null;
  }

  /**
   * Fill template with variables
   */
  fill(template: string, variables: Record<string, string>): string {
    let result = template;
    for (const [key, value] of Object.entries(variables)) {
      result = result.replace(new RegExp(`\\{${key}\\}`, 'g'), value);
    }
    return result;
  }

  /**
   * Get and fill a prompt in one call
   */
  async getAndFill(id: string, variables: Record<string, string>): Promise<string | null> {
    const template = await this.getTemplate(id);
    if (!template) return null;
    return this.fill(template, variables);
  }

  /**
   * List all available prompt IDs
   */
  async listIds(): Promise<string[]> {
    const ids = new Set<string>();

    // Load from Supabase if configured
    if (this.supabaseUrl && this.supabaseKey) {
      try {
        const response = await fetch(
          `${this.supabaseUrl}/rest/v1/nko_prompts?select=id&active=eq.true`,
          {
            headers: {
              apikey: this.supabaseKey,
              Authorization: `Bearer ${this.supabaseKey}`,
            },
          }
        );
        if (response.ok) {
          const rows = await response.json();
          rows.forEach((row: { id: string }) => ids.add(row.id));
        }
      } catch (e) {
        console.warn('Failed to list prompts from Supabase:', e);
      }
    }

    // Load from YAML files
    this.loadAllYamlIds().forEach(id => ids.add(id));

    return Array.from(ids);
  }

  /**
   * List prompts by category
   */
  async listByCategory(category: string): Promise<PromptDefinition[]> {
    const allIds = await this.listIds();
    const prompts: PromptDefinition[] = [];

    for (const id of allIds) {
      const prompt = await this.get(id);
      if (prompt && prompt.category === category) {
        prompts.push(prompt);
      }
    }

    return prompts;
  }

  /**
   * Get all world exploration prompts
   */
  async getWorldPrompts(): Promise<PromptDefinition[]> {
    return this.listByCategory('world_exploration');
  }

  /**
   * Get all frame analysis prompts
   */
  async getFramePrompts(): Promise<PromptDefinition[]> {
    return this.listByCategory('frame_analysis');
  }

  /**
   * Clear the cache
   */
  clearCache(): void {
    this.cache.clear();
  }

  // Private methods

  private async loadFromSupabase(id: string): Promise<PromptDefinition | null> {
    if (!this.supabaseUrl || !this.supabaseKey) return null;

    try {
      const response = await fetch(
        `${this.supabaseUrl}/rest/v1/nko_prompts?id=eq.${id}&active=eq.true`,
        {
          headers: {
            apikey: this.supabaseKey,
            Authorization: `Bearer ${this.supabaseKey}`,
          },
        }
      );

      if (!response.ok) return null;

      const rows = await response.json();
      if (rows.length === 0) return null;

      const row = rows[0];
      return {
        id: row.id,
        name: row.name,
        description: row.description,
        category: row.category,
        cost_tier: row.cost_tier,
        template: row.template,
        output_schema: row.output_schema,
        version: row.version,
        active: row.active,
        tags: row.tags,
      };
    } catch (e) {
      console.warn(`Failed to load prompt ${id} from Supabase:`, e);
      return null;
    }
  }

  private loadFromYaml(id: string): PromptDefinition | null {
    // Search all YAML files
    const files = this.getYamlFiles();
    
    for (const file of files) {
      try {
        const content = fs.readFileSync(file, 'utf-8');
        const data = yaml.load(content) as PromptFile;
        
        if (data.prompts && data.prompts[id]) {
          const prompt = data.prompts[id];
          return {
            ...prompt,
            category: prompt.category || data.category,
            cost_tier: (prompt.cost_tier || data.cost_tier) as 'text_only' | 'multimodal' | 'live',
          };
        }
      } catch (e) {
        console.warn(`Failed to load YAML file ${file}:`, e);
      }
    }

    return null;
  }

  private loadAllYamlIds(): string[] {
    const ids: string[] = [];
    const files = this.getYamlFiles();

    for (const file of files) {
      try {
        const content = fs.readFileSync(file, 'utf-8');
        const data = yaml.load(content) as PromptFile;
        
        if (data.prompts) {
          ids.push(...Object.keys(data.prompts));
        }
      } catch (e) {
        console.warn(`Failed to parse YAML file ${file}:`, e);
      }
    }

    return ids;
  }

  private getYamlFiles(): string[] {
    const files: string[] = [];
    
    try {
      const entries = fs.readdirSync(this.yamlDir, { withFileTypes: true });
      
      for (const entry of entries) {
        if (entry.isFile() && (entry.name.endsWith('.yaml') || entry.name.endsWith('.yml'))) {
          files.push(path.join(this.yamlDir, entry.name));
        }
      }
    } catch (e) {
      console.warn(`Failed to read YAML directory ${this.yamlDir}:`, e);
    }

    return files;
  }
}

// Export singleton instance
export const promptLoader = new PromptLoader();

// Convenience functions
export async function getPrompt(id: string): Promise<string | null> {
  return promptLoader.getTemplate(id);
}

export async function getAndFillPrompt(
  id: string,
  variables: Record<string, string>
): Promise<string | null> {
  return promptLoader.getAndFill(id, variables);
}

